massa-final-state/src/executed_ops.rs

```diff
@@ -22,39 +25,66 @@ use std::ops::Bound::{Excluded, Included};
 
 /// A structure to list and prune previously executed operations
 #[derive(Debug, Default, Clone, PartialEq, Eq)]
-pub struct ExecutedOps(PreHashMap<OperationId, Slot>);
+pub struct ExecutedOps {
+    /// Map of the executed operations
+    ops: PreHashMap<OperationId, Slot>,
+    /// Cumulated hash of the executed operations
+    pub hash: Option<Hash>,
+}
 
 impl ExecutedOps {
     /// returns the number of executed operations
     pub fn len(&self) -> usize {
-        self.0.len()
+        self.ops.len()
     }
 
     /// Check is there is no executed ops
     pub fn is_empty(&self) -> bool {
-        self.0.is_empty()
+        self.ops.is_empty()
     }
 
     /// extends with another ExecutedOps
     pub fn extend(&mut self, other: ExecutedOps) {
-        self.0.extend(other.0);
+        if let Some(other_hash) = other.hash {
+            if let Some(current_hash) = self.hash.as_mut() {
+                *current_hash ^= other_hash;
+            } else {
+                self.hash = Some(other_hash);
+            }
+        }
+        self.ops.extend(other.ops);
```

Comment:
I think you should put this in scrope of the if `if let Some(other_hash) = other.hash` so that ip ensure you don't add executed ops without adding hash. Even if right now it's not possible and you know it because you made the code if someone else make modifications in future it could lead to errors.